### PR TITLE
ENH: add all hooks required to allow transition of pytmc stcmd -> template

### PR DESCRIPTION
**Pull request by klauer**
## Context

Related: https://github.com/pcdshub/pytmc/issues/209
Pairs with: https://github.com/pcdshub/ads-ioc/pull/86

## Changes

* Adds helper commands to `pytmc template`, which can be used in Jinja templates:
    * `generate_records` (create .db and .archive files)
    * `get_plc_by_name`
    * `get_symbols_by_type`
* Adds variables to `pytmc template` environment, which can be used in Jinja templates:
   * `pytmc_version`
* Adds `--macro` option to `pytmc template`
* Fixes some annotations + uncovered/untested functionality
* Need for separate `pytmc stcmd` and `pytmc db` step goes away
   * Either can still exist if desirable, but ads-ioc will only use `pytmc template` for the main build step
* Allows `pytmc template` to read/write multiple templates with parsing a project only once

### CODE REVIEW

File: pytmc/bin/template.py [hunk 1]
```diff
@@ -285,27 +313,64 @@ def get_symbols(plc) -> Generator[parser.Symbol, None, None]:
...
+    """Get symbols for the PLC."""
+    symbols = plc.find(parser.Symbol, recurse=False)
+    return parser.separate_by_classname(symbols)
+
+
 @functools.lru_cache()
-def get_motors(plc) -> list:
-    """Get motor symbols for the PLC (non-pointer DUT_MotionStage)."""
-    symbols = get_symbols(plc)
+def get_motors(plc: parser.Plc) -> list:
```
Comment (ZLLentz) [inline pytmc/bin/template.py:323]: A small handful of these have incomplete return typing and the py39 style.

**Review by ZLLentz — COMMENTED** [review]
It's always really hard to follow the code paths in pytmc. This looks pretty clean, I couldn't focus/follow closely enough to give a more in depth line by line. I'll move to reading through ads-ioc before trying on a few projects.

File: pytmc/bin/template.py (hunk 1 above)
Comment (klauer) [reply pytmc/bin/template.py]: Tried to address these in 5aeb9ce, but this is still a bit of a mess

**Comment by klauer** [comment]
(I mistook the above for approval, sorry @ZLLentz. I'm going to add branch protection now)

**Comment by ZLLentz** [comment]
I'll hit a quick post-review

File: pytmc/bin/template.py [hunk 2]
```diff
@@ -353,29 +412,98 @@ def get_data_types(project) -> List[dict]:
...
+    if write_archive_file:
+        with open(archive_path, "wt") as fp:
+            fp.write('\n'.join(generate_archive_settings(packages)))
+
+    by_pvname = {
+        record.pvname: record
+        for package in packages
+        for record in package.records
+    }
+    return [str(path)], by_pvname
```
Comment (ZLLentz) [inline pytmc/bin/template.py:502]: A bit confusing that the first element of the return tuple is a list of one string- it must match up with an internal interface elsewhere, I guess?

**Comment by ZLLentz** [comment]
Everything looks fine- I didn't follow all the paths but nothing super weird, I'm going to focus on other PRs

File: pytmc/bin/template.py (hunk 2 above)
Comment (klauer) [reply pytmc/bin/template.py:502]: It is a wonky return type now that you mention it. It would have been better to return a structured type or a dict or something.
